backend/tournesol/views/subsamples.py

```diff
@@ -0,0 +1,85 @@
+"""
+The sub-samples API return a sub-sample of entities compared by a user.
+
+A sub-sample is a small finite number of compared entities, ranked by the
+individual score computed for the poll's main criterion.
+"""
+import random
+
+from django.db.models import Window
+from django.db.models.functions import Ntile
+from drf_spectacular.types import OpenApiTypes
+from drf_spectacular.utils import OpenApiParameter, extend_schema
+from rest_framework import generics
+from rest_framework.permissions import IsAuthenticated
+
+from tournesol.models import ContributorRating
+from tournesol.serializers.subsample import SubSampleSerializer
+from tournesol.views.ratings import ContributorRatingQuerysetMixin
+
+# The number of ranked group used to divide the rated entities.
+DEFAULT_BUCKET_SIZE = 20
+
+
+class SubSamplesQuerysetMixin(ContributorRatingQuerysetMixin):
+    def get_queryset(self):
+        poll = self.poll_from_url
+
+        try:
+            sub_sample_size = int(self.request.query_params["ntile"])
+        except KeyError:
+            sub_sample_size = DEFAULT_BUCKET_SIZE
+
+        qst = (
+            ContributorRating.objects.annotate_n_comparisons()
+            .annotate_collective_score()
+            .annotate_individual_score(poll=poll)
+            .prefetch_related(self.get_prefetch_entity_config())
+            .filter(
+                poll=poll,
+                user=self.request.user,
+                criteria_scores__criteria=poll.main_criteria
+            )
+            .annotate(bucket=Window(
+                expression=Ntile(sub_sample_size),
+                order_by="-criteria_scores__score"
+            ))
+        )
+
+        sub_sample = []
+        buckets = len(qst)
+
+        for i in range(min(buckets, sub_sample_size)):
+            sub_sample.append(
+                random.choice([rating for rating in qst if rating.bucket == i + 1])  # nosec
```

Comment:
All ratings are fetched with all annotations from prefetched global and individual scores. That's suboptimal, but maybe it works perfectly fine and does not need to be optimized?

Proposal 1:
Don't do the bucketing on the database side, fetch everything from a sorted qs, use np.array_split for the bucketing. I expect it would be faster than the for loop you are using that then goes through the whole list to group each bucket together.
```
sorted_ratings = list(qst)  # assuming the qst has order_by
sub_sample_size = min(sub_sample_size, len(sorted_ratings)
sub_sample = [random.choice(bucket) for bucket in np.array_split(sorted_ratings, sub_sample_size)]
```

Proposal 2:
Make a first query for only the `uid`, fetch everything, sample from buckets, then do a second query with the selected list of uids that includes all the metadatas.

Reply:
Good idea!

The for loop seems to be the most efficient part of the code. It's the SQL query, especially the annotation `annotate_n_comparisons`, that severely hits the performances. The SQL query can take up to 1 second for a user having compared ~3000 entities.

I'll try the proposal 2, to see how it improves the performance. And maybe the proposal 1, even if I think it will have a positive impact only if the SQL window function is not optimized.

Reply:
I implemented the proposal 2. It greatly improves the performance (see: https://github.com/tournesol-app/tournesol/pull/1765/commits/cdee9f41450c6c6ee8045c0feb85efd64425bfd0).

Currently the buckets are made by the database and the ratings are selected with Python. It currently takes ~ 5 milliseconds to perform the SQL query with the window function, and according to the Python `time` module ~ 5 more milliseconds to iterate over 2300 ratings.

If we need to improve this we have two strategies:
- create the buckets and select the ratings in one SQL query / no Python
- select all ratings by SQL and create the buckets and select the entities with Python only

I haven't tried any of those strategies yet as 10 ms seems to be good enough. Let me know what do you think. We can still try by curiosity : )